pkg/globalmanager/controllers/objectsearch/objectsearchservice.go

```diff
@@ -0,0 +1,614 @@
+}
+
+// sync will sync the objectsearchservice with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (c *Controller) sync(key string) (bool, error) {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing objectsearch service %q (%v)", key, time.Since(startTime))
+	}()
+
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return false, err
+	}
+	if len(ns) == 0 || len(name) == 0 {
+		return false, fmt.Errorf("invalid objectsearch service key %q: either namespace or name is missing", key)
+	}
+	sharedService, err := c.serviceLister.ObjectSearchServices(ns).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.V(4).Infof("ObjectSearchService has been deleted: %v", key)
+			return true, nil
+		}
+		return false, err
+	}
+
+	service := *sharedService
+
+	// if service was finished previously, we don't want to redo the termination.
+	if isServiceFinished(&service) {
+		return true, nil
+	}
+
+	// set kind for service in case that the kind is None.
+	// more details at https://github.com/kubernetes/kubernetes/issues/3030
+	service.SetGroupVersionKind(Kind)
+
+	selectorPods, _ := runtime.GenerateWorkerSelector(&service, objectSearchTrackingWorker)
+	selectorDeployments, _ := runtime.GenerateSelector(&service)
+	pods, err := c.podStore.Pods(service.Namespace).List(selectorPods)
+	deployments, err := c.deploymentsLister.Deployments(service.Namespace).List(selectorDeployments)
+	if err != nil {
+		return false, err
+	}
+
+	latestConditionLen := len(service.Status.Conditions)
+
+	var podFailed int32 = 0
+	var deploymentFailed int32 = 0
+
+	// neededPodCounts indicates the num of tracking worker pods should be created successfully in a objectsearch service currently.
+	// neededDeploymentCounts indicates the num of deployments should be created successfully in a objectsearch service currently,
+	// and one deployment is for userWorker and the other deployment is for reidWorkers.
+	var neededPodCounts int32 = int32(len(service.Spec.TrackingWorkers))
+	var neededDeploymentCounts int32 = 2
+
+	activePods := runtime.CalcActivePodCount(pods)
+	activeDeployments := runtime.CalcActiveDeploymentCount(deployments)
+
+	if service.Status.StartTime == nil {
+		now := metav1.Now()
+		service.Status.StartTime = &now
+	} else {
+		podFailed = neededPodCounts - activePods
+		deploymentFailed = neededDeploymentCounts - activeDeployments
+	}
+
+	var manageServiceErr error
+	serviceFailed := false
+
+	var latestConditionType sednav1.ObjectSearchServiceConditionType = ""
+
+	// get the latest condition type
+	// based on that condition updated is appended, not inserted.
+	jobConditions := service.Status.Conditions
+	if len(jobConditions) > 0 {
+		latestConditionType = (jobConditions)[len(jobConditions)-1].Type
+	}
+
+	var newCondtionType sednav1.ObjectSearchServiceConditionType
+	var reason string
+	var message string
+
+	if podFailed > 0 || deploymentFailed > 0 {
```

Comment:
might can use switch-case instead of if-else

Reply:
Changed to  switch-case.